serde_codegen/src/ser.rs

```diff
@@ -226,9 +227,10 @@ fn serialize_tuple_struct(
 
     let type_name = name_expr(builder, item_attrs.name());
     let len = serialize_stmts.len();
+    let let_mut = mut_if_nonzero(cx, len);
 
     quote_block!(cx, {
-        let mut state = try!(_serializer.serialize_tuple_struct($type_name, $len));
+        let $let_mut state = try!(_serializer.serialize_tuple_struct($type_name, $len));
         $serialize_stmts
```

Comment:
@nox I think I would rather move in the opposite direction, of making our generated code more straightforward where that can be accomplished easily. See as examples #479 and #538. There are a few reasons that is valuable:

1. The most straightforward code will give us (and others) the best experience when the generated code needs to be debugged.
2. The generated code is valuable as a template for implementing Serialize and Deserialize manually with minor modifications.
3. Convoluted generated code gives people the impression that Serde is hard or Serde codegen is magic and impossible to understand, scaring people away from implementing Serialize and Deserialize manually where necessary.

I'm not saying we need to make the generated code beautiful at the expense of maintainability of serde_codegen, but there is a middle ground and I don't think rolling `let mut state = state` into `$serialize_stmts` is the right tradeoff in this case. Thanks for the idea though!